R/get_ref.R

```diff
@@ -150,8 +150,19 @@ get_ref_min.remote_ref_github <- function(remote_ref, op = "", op_ver = "") {
     }
   }
 
-  new_ref <- sprintf("%s=%s/%s%s", remote_ref$package, remote_ref$username, remote_ref$repo, ref_suffix) # @TODO
-  pkgdepends::parse_pkg_ref(new_ref)
+  new_ref <- sprintf(
+    "%s=%s/%s%s",
+    remote_ref$package,
+    remote_ref$username,
+    remote_ref$repo,
+    ref_suffix
+  )
+  result <- pkgdepends::parse_pkg_ref(new_ref)
+
+  # Needs to restore original ref as it needs it to solve the pkg tree. With error:
+  #  Can't install dependency <original_ref> (>= <version>)
+  result$ref <- remote_ref$ref
+  result
```

Comment:
I am very interested in reasoning here. Can you please elaborate more?
I am a little concerned because this is going to be the first break of self-consistency of this object which is returned(!) (out of exported! fun) so we don't really know who and how it is used. From what I recall, sometimes we use `ref` element and sometimes we use other like `version` etc. With this change we need to be very very careful. And I would like to avoid such traps if possible

Reply:
The previous behaviour was breaking consistency as it changed the original `ref` from the remotes with the output of this function.

It replaces `insightsengineering/dunlin@*release` with `dunlin=insightsengineering/dunlin@v0.1.3`. _(screenshot below with the error and `get_resolution()`)_

As a result, when it is trying to solve the dependencies it detects that the `...@*release` ref cannot be installed.

An alternate solution would be for us to populate the `version`, `commitish` and any other field fields manually. However, I really like using `pkgdepends::parse_pkg_ref` for this as it reduces possible maintenance in the future.

![image](https://github.com/insightsengineering/verdepcheck/assets/211358/a1313cd3-c1e6-4e72-81f8-3ab9b6c8b7c7)

![image](https://github.com/insightsengineering/verdepcheck/assets/211358/53ac58cc-fa39-47ba-a518-0135ea9c9234)

Reply:
I've been thinking about this, testing and debugging and trying to understand the whole context better and design a good solution.

🗓️ I've sent an invite for us to talk on early next week (Monday?) about this.

#### Addressing your last point first:

> it looks like an installation error as opposed to dependency resolve conflicts

It fails to install as the resolution had no candidate for  the `.../...@*release` as the ref name was replaced by `dunlin=.../dunlin@v0.1.3`.

I'm very confident that it's not a token or installation problem as the message appears to indicate.

#### The forced assignment after `pkgdepends::parse_pkg_ref(new_ref)`

⚠️ The comments bellow are only for the `min_isolated` strategy

I think the problem we have with `Remotes` is a bit wider than we realised, there are currently 3 issues _(maybe more as I found a weird big also on `min_cohort`)_

1. Your PR on https://github.com/r-lib/pkgdepends/pull/321
1. Can't install dependency `.../...@*release`
    * Proposed fix that started this discussion
1. Min version is not respected by packages that are described on Remotes dependency that also exist on CRAN
    * For example, `formatters` _(as a primary dependency)_ will install `0.5.9003` instead of `0.5.0`

To help me understand better the whole process I made a diagram for this strategy's technical workflow _(see at the end)_

### 💡 Idea for solution 

I think the solution should be targetted when the IP is being prepared _(during first call on the diagram)

In particular, when the `Need/Config/verdepcheck` resolves to a particular version or to CRAN, it will modify/remove the corresponding remote.

* If it resolves to a particular tag, it changes the `Remotes` to that version instead of `@*release`. If the package is resolved to CRAN then it will remove the corresponding `Remotes`.
* (alternative) Solve this downstream _(second call on the diagram)_
  * For issue 2. using my first fix
  * For issue 3. implement logic that detects this scenario and uses a similar solution

![min_isolated drawio(4)](https://github.com/insightsengineering/verdepcheck/assets/211358/2438eb9e-9830-4997-b6cf-e43d6526ac2e)

Reply:
## Summary:

@pawelru I've Implemented the solution we discussed on our meeting today and my initial tests show it's:

* ✅ working very well with `{chevron}` _(pkg where some of these problems were materializing)_
* ❌ The local edge case we talked about confirms our hypothesis with the encapsulation scenario
    * When a primary dependency is included in the dependency tree with a `Remotes` _(resolving to `main` branch)_

Here are the changes: [6079c3c4..caea5c37](https://github.com/insightsengineering/verdepcheck/pull/26/files/6079c3c439b7ac316da7093d2ef0220cbad1bd5c..caea5c374e923354a0c3baea9336cd5067ca54b1) -- Changes applied to all strategies, but `max`.

### On `chevron` it :

* Resolves problematic package `{dunlin}` without any issues _(installing the expected tag)_
* Corrects another bug _(that we didn't discuss)_ with `{formatters}`, installing the "min_isolated" version instead of the latest from main.

### Local edge case (`{test.repo}`)

_(`Imports: rtables, tern and averissimo/teal.logger`)_

`averissimo/teal.logger` is a non-CRAN pkg with custom `Remotes` that includes `formatters`, which is one of the primary dependencies of `{test.repo}`

It tries to install `formatters 0.5.0.9004` instead of `0.5.0` and it validates our hypothesis with the encapsulation problem.

⚠️ note: all packages that are in CRAN will have GitHub `Remotes` ignored and squashed in favor of "ppm", so this example with only `{tern}` and `{rtables}` doesn't have any issues.

![image](https://github.com/insightsengineering/verdepcheck/assets/211358/a4decf95-192a-4d74-af6c-c6b2c977d4f5)

![image](https://github.com/insightsengineering/verdepcheck/assets/211358/e52542ad-8adb-464a-a2ed-f5cacd516c3c)